pypi_mirror.py

```diff
@@ -58,34 +61,94 @@ def are_all_pkgs_downloaded(local_index_path):
 
 
 def update_mirror():
-    """update local mirror with recent events from pypi"""
+    """
+    Handles changelog events as follows:
+    - 'new', 'add', 'create', 'update', 'docupdate': Downloads/updates the pkg
+    - 'remove': Deletes the package dir
+    - 'rename': Deletes the old repository, then downloads/updates the renamed pkg
+    """
     with open(last_serial_path, 'r') as f:
         serial = int(f.readline().split(":")[1].strip())
 
-    packages, current_serial = set(), client.changelog_last_serial()
-    # changelog_since_serial(serial) return only 50k changes per query and the returned numbers 
+    current_serial = client.changelog_last_serial()
+
+    pkgs_to_download, pkgs_to_remove, last_serial = get_changes(serial, current_serial)
+
+    print(f"Removing {len(pkgs_to_remove)}, Adding {len(pkgs_to_download)} pkgs" )
+
+    handle_removals(pkgs_to_remove)
+    handle_downloads(pkgs_to_download)
+    
+    if pkgs_to_download:
+        artifacts_metadata = generate_updated_csv(pkgs_to_download, os.path.join(mirror_dir, 'mirror_metadata.csv'))
+        send_add_requests(artifacts_metadata)
```

Comment:
At the very least, we should check whether RSTUF successfully ran all the Tasks.